src/cpp/grid.hpp

```diff
@@ -0,0 +1,133 @@
+#ifndef GRID_H
+#define GRID_H
+
+#include <limits>
+#include <string>
+#include <vector>
+
+#include "geom.hpp"
+#include "transform.hpp"
+#include "utils.hpp"
+
+/// Structure defining a raster grid.
+struct Grid {
+  /// number of columns
+  size_t ncols;
+  /// number of rows
+  size_t nrows;
+  /// grid to world transform - provided (derived from x, y offset and cellsize)
+  Affine grid_to_world;
+  /// world to grid transform - calculated
+  Affine world_to_grid;
+  /// 1D vector of doubles storing the data (conceptually a 2D grid)
+  std::vector<double> data;
+
+  Grid()
+      : ncols{0}, nrows{0},
+        grid_to_world{Affine()}, data{std::vector<double>()} {
+    world_to_grid = ~grid_to_world;
+  };
+
+  Grid(size_t ncols, size_t nrows, Affine grid_to_world)
+      : ncols{ncols}, nrows{nrows},
+        grid_to_world{grid_to_world}, data{std::vector<double>()} {
+    world_to_grid = ~grid_to_world;
+  };
+
+  Grid(size_t ncols, size_t nrows, Affine grid_to_world,
+       std::vector<double> data)
+      : ncols{ncols}, nrows{nrows}, grid_to_world{grid_to_world}, data{data} {
+    world_to_grid = ~grid_to_world;
+  };
+
+  /// Calculate hashed index in raster.
+  int cellIndex(const geometry::Vec2<double> p) const {
+    auto offset = cellIndices(p);
+    return offset.x + offset.y * ncols;
+  }
+
+  /// Recover i, j index in raster.
+  geometry::Vec2<int>
+  cellIndices(const geometry::Vec2<double> p,
+              double epsilon = std::numeric_limits<double>::epsilon()) const {
+    auto offset =
+        world_to_grid * (p + geometry::Vec2<double>(epsilon, epsilon));
```

Comment:
maybe a comment explaining the purpose of epsilon? It's a bit mysterious to me why it is needed. But could be obvious for domain expert?

Reply:
Added a note - this is anticipating precision problems that could come through the transform, but I'm struggling to write an actual test case. Could drop it until we understand there really is a problem that it solves?

Reply:
Yes of course, adding the note and TODO is good enough!